sdk/src/Core/Amazon.Runtime/Credentials/Internal/IdentityResolvers/DefaultAWSCredentialsIdentityResolver.cs

```diff
@@ -0,0 +1,191 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+using Amazon.Runtime.CredentialManagement;
+using Amazon.Runtime.Internal.Util;
+using Smithy.Identity.Abstractions;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Security;
+using System.Threading;
+
+namespace Amazon.Runtime.Credentials.Internal.IdentityResolvers
+{
+    /// <summary>
+    /// A resolver that provides an AWSCredentials identity. 
+    /// This implementation is used when scheme ID is sigv4 or sigv4a, 
+    /// </summary>
+    public class DefaultAWSCredentialsIdentityResolver : IIdentityResolver
+    {
+        internal const string AWS_PROFILE_ENVIRONMENT_VARIABLE = "AWS_PROFILE";
+        internal const string DEFAULT_PROFILE_NAME = "default";
+
+        private static ReaderWriterLockSlim cachedCredentialsLock = new ReaderWriterLockSlim();
+        private delegate AWSCredentials CredentialsGenerator();
+        private AWSCredentials cachedCredentials;
+        private List<CredentialsGenerator> CredentialsGenerators { get; set; }
+        private readonly CredentialProfileStoreChain credentialProfileChain = new CredentialProfileStoreChain();
+
+        private DefaultAWSCredentialsIdentityResolver()
+        {
+            try
+            {
+                cachedCredentialsLock.EnterWriteLock();
+                cachedCredentials = null;
+                CredentialsGenerators = new List<CredentialsGenerator>
+                {
+#if BCL
+                () => new AppConfigAWSCredentials(), // Test explicit keys/profile name first.
+#endif
+                    () => AssumeRoleWithWebIdentityCredentials.FromEnvironmentVariables(),
+                    () => new EnvironmentVariablesAWSCredentials(), // Look for credentials set in environment vars.
+                    () => GetAWSCredentials(credentialProfileChain),
+                    () => ContainerEC2CredentialsWrapper(), // either get ECS / EKS credentials or instance profile credentials
+                };
+            }
+            finally
+            {
+                cachedCredentialsLock.ExitWriteLock();
+            }
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "We need to catch all exceptions to be able to move the the next generator.")]
+        public BaseIdentity ResolveIdentity()
+        {
+            try
+            {
+                cachedCredentialsLock.EnterReadLock();
```

Comment:
I will need to double check if locking is required with this singleton implementation, probably I will only need to add locking when creating the singleton instance.

Reply:
nvm, I don't think we should remove the locking.